diff -up openssl-1.1.1-pre8/include/openssl/opensslv.h.version-override openssl-1.1.1-pre8/include/openssl/opensslv.h
--- openssl-1.1.1-pre8/include/openssl/opensslv.h.version-override	2018-06-20 16:48:13.000000000 +0200
+++ openssl-1.1.1-pre8/include/openssl/opensslv.h	2018-07-25 11:03:29.297543392 +0200
@@ -39,8 +39,8 @@ extern "C" {
  * (Prior to 0.9.5a beta1, a different scheme was used: MMNNFFRBB for
  *  major minor fix final patch/beta)
  */
-# define OPENSSL_VERSION_NUMBER  0x10101008L
-# define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.1-pre8 (beta) 20 Jun 2018"
+# define OPENSSL_VERSION_NUMBER  0x1010100fL
+# define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.1-pre8 (beta) FIPS 20 Jun 2018"
 
 /*-
  * The macros below are to be used for shared library (.so, .dll, ...)
