diff -up openssl-1.1.1/apps/speed.c.coverity openssl-1.1.1/apps/speed.c
--- openssl-1.1.1/apps/speed.c.coverity	2018-10-09 16:32:44.912051009 +0200
+++ openssl-1.1.1/apps/speed.c	2018-10-09 16:29:55.518851544 +0200
@@ -2852,7 +2852,7 @@ int speed_main(int argc, char **argv)
 
             if (rsa_count <= 1) {
                 /* if longer than 10s, don't do any more */
-                for (testnum++; testnum < EC_NUM; testnum++)
+                for (testnum++; testnum < ECDSA_NUM; testnum++)
                     ecdsa_doit[testnum] = 0;
             }
         }
