diff -up openssl-1.0.2h/ssl/ssl_lib.c.dtls-bad-ver openssl-1.0.2h/ssl/ssl_lib.c
--- openssl-1.0.2h/ssl/ssl_lib.c.dtls-bad-ver	2016-05-03 18:06:45.000000000 +0200
+++ openssl-1.0.2h/ssl/ssl_lib.c	2016-08-10 13:47:33.995530954 +0200
@@ -1828,7 +1828,7 @@ int SSL_export_keying_material(SSL *s, u
                                const unsigned char *p, size_t plen,
                                int use_context)
 {
-    if (s->version < TLS1_VERSION)
+    if (s->version < TLS1_VERSION && s->version != DTLS1_BAD_VER)
         return -1;
 
     return s->method->ssl3_enc->export_keying_material(s, out, olen, label,
